cpp/src/arrow/array/builder_nested.cc

```diff
@@ -135,17 +140,28 @@ void StructBuilder::Reset() {
 
 Status StructBuilder::FinishInternal(std::shared_ptr<ArrayData>* out) {
   RETURN_NOT_OK(TrimBuffer(BitUtil::BytesForBits(length_), null_bitmap_.get()));
-  *out = ArrayData::Make(type_, length_, {null_bitmap_}, null_count_);
 
-  (*out)->child_data.resize(children_.size());
+  std::vector<std::shared_ptr<ArrayData>> child_data(children_.size());
   for (size_t i = 0; i < children_.size(); ++i) {
     if (length_ == 0) {
       // Try to make sure the child buffers are initialized
       RETURN_NOT_OK(children_[i]->Resize(0));
     }
-    RETURN_NOT_OK(children_[i]->FinishInternal(&(*out)->child_data[i]));
+    RETURN_NOT_OK(children_[i]->FinishInternal(&child_data[i]));
   }
 
+  // If the type has not been specified, infer it.
```

Comment:
Yeah, this needs some more discussion I think, see my above comment.

Reply:
I still think it would be better to do this inference in the `StructBuilder` constructor, just like we do for `ListBuilder`. Is there a reason not to do it?